Fix read_avro() skip_rows and num_rows.

tpn opened the pull request:
## Description

This PR fixes the avro reader (`cudf.read_avro()`) such that it honors the values passed to the `skip_rows` and `num_rows` parameters.  In implementing this new logic, we also revamp the reader's ability to handle multi-block avro files, which we also test extensively with a new `test_avro_reader_multiblock()` test that features some 1300 permutations of various block size combinations.

Closes #6529.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

tpn commented:
Can someone from NVIDIA add the `bugfix,non-breaking` labels?  I don't have permission.  Thanks!

galipremsagar commented:
/ok to test

galipremsagar commented:
/ok to test

tpn commented:
Looks like the only failure is due to s3 flakiness:
```
[rapids-download-from-s3] Downloading and decompressing s3://rapids-downloads/ci/cudf/pull-request/12912/8bf4378/cudf_conda_cpp_cuda11_x86_64.tar.gz into /tmp/cpp_channel
download failed: s3://rapids-downloads/ci/cudf/pull-request/12912/8bf4378/cudf_conda_cpp_cuda11_x86_64.tar.gz to - An error occurred while reading from response stream: ("Connection broken: ConnectionResetError(104, 'Connection reset by peer')", ConnectionResetError(104, 'Connection reset by peer'))
```

tpn commented:
Just rebased, friendly request for another `/ok to test`.  (Last test failed with a weird unrelated Java error.)

galipremsagar commented:
@tpn could you rebase this PR with `branch-23.06`?

tpn commented:
On it!

tpn commented on python/cudf/cudf/tests/test_avro_reader_fastavro_integration.py:
```diff
@@ -25,12 +25,17 @@
+    schema: dict,
+    records: list,
+    num_rows: Optional[int] = None,
+    skip_rows: Optional[int] = None,
```
If I had a good reason, I can't remember it.  Fixed and pushed.

ttnghia commented:
/ok to test

ttnghia commented on cpp/src/io/avro/avro.hpp:
```diff
@@ -62,13 +62,14 @@ struct column_desc {
+  uint64_t sync_marker[2]  = {0, 0};
+  size_t metadata_size     = 0;
+  size_t total_data_size   = 0;
+  size_type num_rows       = 0;
+  size_type skip_rows      = 0;
+  size_type total_num_rows = 0;
```
Here we have `num_rows` and `total_num_rows` which are confusing. Can you add comments clarifying what is their difference please?

tpn replied:
Roger, will do.

ttnghia commented on cpp/src/io/avro/avro.cpp:
```diff
@@ -96,19 +96,21 @@ bool container::parse(file_metadata* md, size_t max_num_rows, size_t first_row)
       total_object_count += object_count;
       if (!md->block_list.size()) {
-        md->skip_rows = static_cast<uint32_t>(first_row);
+        md->skip_rows = first_row;
```
This was assigned in line 99. So maybe line 99 or this line is incorrect.

tpn replied:
Yeah the initial `md->skip_rows = first_row;` assignment that I added was key to actually solving the bug of skip rows being ignored.

And this subsequent assignment you've highlighted should be redundant, as it's within the `if (!md->block_list.size())` scope, so it should only be hit on the first block.

i.e., this should hold:

```diff
-        md->skip_rows = first_row;
+        CUDF_EXPECTS(md->skip_rows == first_row, "Invariant check failed: md->skip_rows != first_row");
```

However, a few lines up, the whole thing is guarded by `if (object_count > first_row) `... then `else { first_row -= object_count }`.

That seems problematic.  What if the first block has less than or equal rows to the requested number of rows to be skipped? (i.e. the value of `first_rows` at that comparison).  Also, in that case, the block doesn't even get added to the `block_list`, which doesn't seem correct.

I'm going to review a bit more.  At the very least I think there needs to be some tests that exercise the multiple block logic.

tpn replied:
(The suggested code isn't correct, but I think the analysis is on point.)

tpn replied:
I did some digging and thankfully `fastavro` has support for writing multiple blocks via a `sync_interval=` param to the writer.  I'm going to write a few tests to exercise this logic now.  Then we can reconvene and review the proposed C++ changes.

wence- reviewed (approved):
Python changes look fine I think, thanks.

tpn commented on python/cudf/cudf/tests/test_avro_reader_fastavro_integration.py:
```diff
@@ -25,12 +25,17 @@
+    skip_rows: Optional[int] = None,
+    num_rows: Optional[int] = None,
+) -> cudf.DataFrame:
     schema = [] if schema is None else fastavro.parse_schema(schema)
```
Ah, you're right, good catch.  I added that support prior to realizing I'd need to construct the data frames differently in the new tests.  I'm indifferent as to whether or not it should stay or be removed -- any preference on your end?

wence- replied:
Let's remove it in the interest of helping any future blamers.

tpn replied:
Will do.

tpn commented on python/cudf/cudf/tests/test_avro_reader_fastavro_integration.py:
```diff
@@ -444,3 +449,97 @@ def test_alltypes_plain_avro():
+
+@pytest.mark.parametrize(
+    "total_rows_and_num_rows",
+    [(1, 1), (10, 5), (100, 10), (10, 10), (10, 0), (10, 9), (10, 1)],
```
I think most of them yielded crashes in some way or another when I was developing the fix :-)

But it's basically just testing boundary conditions.

wence- replied:
Perhaps if you can recall, you can comment the cases with the boundary condition that is being tested? But if not, no worries.

vuule commented on cpp/src/io/avro/avro.cpp:
```diff
@@ -92,23 +94,38 @@ bool container::parse(file_metadata* md, size_t max_num_rows, size_t first_row)
 
   md->metadata_size  = m_cur - m_base;
-  md->skip_rows      = 0;
+  md->skip_rows      = first_row;
```
I don't understand this fix. At a glance, the old logic makes sense to me: iterate over blocks and reduce the number of rows to skip until we don't need to skip all rows in the current block. Record the number of rows to skip in the first block that is included in block_list (which only has block that we at least partially read).
What was incorrect in this logic?

tpn replied:
I'm going to write some multi-block tests and come back to this (very reasonable) question.  I am re-reading the code and I'm inclined to agree -- perhaps the only fix that's needed is this bit that was at the end of the while loop:

```cpp

  if ((max_num_rows <= 0) || (max_num_rows > total_object_count)) {
    md->num_rows = total_object_count;
  } else {
    md->num_rows = max_num_rows;
  }
```

tpn replied:
Addressed in the most recent commit.

tpn commented:
I've just pushed a commit with the `md->skip_rows = first_row;` reverted, plus a new multi-block test.  However, I can't seem to get branch-23.06 to build, all of the string equality lines in `datetime.pyx` are bombing out with this:

```
Error compiling Cython file:
------------------------------------------------------------
...
            )
        elif field == "microsecond":
            c_result = move(
                libcudf_datetime.extract_microsecond_fraction(col_view)
            )
        elif field == "nanosecond":
                     ^
------------------------------------------------------------

/home/trent/src/cudf/python/cudf/cudf/_lib/datetime.pyx:64:22: Operation not allowed without gil
```

(I created a new environment for 23.06: `mamba env create -n cudf2306 --file conda/environments/all_cuda-118_arch-x86_64.yaml`.)

23.04 seems to build fine, still.  I'll apply this patch to that branch for now so I can at least verify the multi-block logic works.  An `/ok to test` would be great just so I can see if the builders hit the same issue.

tpn commented:
I'm still working on more multiblock tests.  I think I've found some combinations that are exposing some bugs in the logic.  I'll continue work on it tomorrow, but only have the morning available, so it likely won't be until early next week when I have another update to push.

tpn commented:
Update: I'm still working on this.  The new multi-block tests I added exposed a whole host of corner-case failures in both the `skip_rows` *and* `num_rows` logic.  I was hoping to have everything finished today, but that is unlikely at this point.

tpn commented:
Good news, looks like I've got all the multi-block issues worked out.  Quite a lot of stuff had to change.  I'm working on cleaning up the patches now and should be ready to commit within the next day or two.

vuule commented:
> Good news, looks like I've got all the multi-block issues worked out. Quite a lot of stuff had to change. I'm working on cleaning up the patches now and should be ready to commit within the next day or two.

Thank you for digging into this. We definitely have a gap in testing with large input (as in, we don't have tests with large inputs :D).

tpn commented:
Just pushed the latest changes with multi-block fixes.  Friendly request for some more reviewer eyes, plus another `/ok to test`.  Thanks!

vuule commented:
/ok to test

tpn commented:
Just pushed a rebase... not sure what happened with the earlier one this morning.  Friendly request for another `/ok to test` please :-)

vuule commented:
/ok to test

ttnghia commented:
/ok to test

tpn commented:
Oof, the CentOS 7 failures look pathological... device error perhaps?  They don't look related to this change.

galipremsagar commented:
> Oof, the CentOS 7 failures look pathological... device error perhaps? They don't look related to this change.

Flaky OOMs

galipremsagar commented:
/ok to test

vuule commented on cpp/src/io/avro/avro_gpu.cu:
```diff
@@ -307,42 +352,64 @@ __global__ void __launch_bounds__(num_warps * 32, 2)
   cur            = avro_data + blk->offset;
   end            = cur + blk->size;
-  while (rows_remaining > 0 && cur < end) {
+  first_row      = blk->first_row + blk->row_offset;
+  cur_row        = blk->row_offset;
+  end_row        = first_row + blk->num_rows;
+  rows_remaining = blk->num_rows;
```
Please declare at the point of initialization (if possible). We're trying to move away from the C-style declaration at the top of each function so consistency with the current code is not recommended in this regard :)

vuule commented on cpp/src/io/avro/avro_gpu.cu:
```diff
@@ -283,17 +329,16 @@ __global__ void __launch_bounds__(num_warps * 32, 2)
   __shared__ __align__(8) schemadesc_s g_shared_schema[max_shared_schema_len];
   __shared__ __align__(8) block_desc_s blk_g[num_warps];
 
+  bool skipped_row;
```
I don't see this used outside of the block with `avro_decode_row`. Can it be declared there?

vuule commented on cpp/src/io/avro/avro_gpu.cu:
```diff
@@ -57,24 +57,59 @@ static inline int64_t __device__ avro_decode_zigzag_varint(const uint8_t*& cur,
+  // processing multiple blocks, i.e. this block could only have 10 rows, but
+  // it's the 3rd block (where each block has 10 rows), so we need to write to
+  // the 30th row in the destination array.
+  const ptrdiff_t row_idx =
```
Maybe `dst_row`?

vuule commented on cpp/src/io/avro/avro_gpu.cu:
```diff
@@ -107,22 +142,27 @@ avro_decode_row(schemadesc_s const* schema,
+        if (dataptr != nullptr && row_idx >= 0) {
+          atomicAnd(static_cast<uint32_t*>(dataptr) + (row_idx >> 5), ~(1 << (row_idx & 0x1f)));
           atomicAdd(&schema_g[i].count, 1);
+          *skipped_row = false;
```
Could we centralize changes to `skipped_row`? Is perhaps `dataptr == nullptr || row_idx < 0` the only condition we need to set `skipped_row` to `true`?

vuule reviewed (commented):
Partial review
Great changes, got some small suggestions so far.

wence- commented on python/cudf/cudf/tests/test_avro_reader_fastavro_integration.py:
```diff
@@ -447,3 +450,213 @@ def test_alltypes_plain_avro():
+    ids=["use_sync_interval", "ignore_sync_interval"],
+)
+@pytest.mark.parametrize("codec", ["null", "deflate", "snappy"])
+@pytest.mark.parametrize(
+    "total_rows_and_num_rows_and_skip_rows_and_rows_per_block",
+    total_rows_and_num_rows_and_skip_rows_and_rows_per_block,
+    ids=total_rows_and_num_rows_and_skip_rows_and_rows_per_block_ids,
+)
```
I would probably just turn this into a fixture:
```python
def name_fixture(param):
    (total_rows, num_rows, skip_rows, sync_interval) = param
    # Always reporting num-rows, and using magic f-string `{foo=}` formatting to produce foo=value_of_foo
    return f"{total_rows=}-{num_rows=}-{skip_rows=}-{sync_interval=}"

@pytest.fixture(params=[list_of_things_here], ids=name_fixture)
def total_rows_and...(request):
    return request.param
```

wence- reviewed (commented):
A few minor suggestions in the python tests.

shwina commented on python/cudf/cudf/tests/test_avro_reader_fastavro_integration.py:
```diff
@@ -447,3 +450,213 @@ def test_alltypes_plain_avro():
+        bytes_per_row = 4 if dtype == "float32" else 8
+        avro_type = "float" if dtype == "float32" else "double"
+
+        # We don't use rand_dataframe() here, because it increases the
+        # execution time of each test by a factor of 10 or more (it appears
+        # to use a very costly approach to generating random data).
```
As noted in https://github.com/rapidsai/cudf/issues/13128, you can try to use `use_threads=False` with `rand_dataframe` to see if that is any faster.

tpn commented:
With the latest rebase, I believe I've addressed all PR feedback items.  Let me know if I've missed anything.  Should be good for another `/ok to test`.

vuule commented:
/ok to test

vuule commented:
/ok to test

vuule commented on cpp/src/io/avro/avro.cpp:
```diff
@@ -92,36 +94,113 @@ bool container::parse(file_metadata* md, size_t max_num_rows, size_t first_row)
+    sync_marker[1] = get_raw<uint64_t>();
+    valid_sync_markers =
+      ((sync_marker[0] == md->sync_marker[0]) && (sync_marker[1] == md->sync_marker[1]));
+    if (!valid_sync_markers) { return false; }
```
Can we `CUDF_EXPECTS(valid_sync_markers, ...)` here, as the comment (maybe) suggests?

vuule commented on cpp/src/io/avro/avro.cpp:
```diff
@@ -92,36 +94,113 @@ bool container::parse(file_metadata* md, size_t max_num_rows, size_t first_row)
+    // Abort on terminal conditions.  We keep these as separate lines instead of
+    // combining them into a single if in order to facilitate setting specific
+    // line breakpoints in the debugger.
+    if (block_size <= 0) { return false; }
+    if (object_count <= 0) { return false; }
+    if (next_end > m_end) { return false; }
```
CUDF_EXPECTS?
We can leave the existing exists as they are, but it would be nice to start switching to exceptions in the new code.
Edit: the change is already a huge improvement :+1:

vuule commented on cpp/src/io/avro/avro.cpp:
```diff
@@ -67,8 +67,10 @@ std::string container::get_encoded()
 {
   constexpr uint32_t avro_magic = (('O' << 0) | ('b' << 8) | ('j' << 16) | (0x01 << 24));
-  uint32_t sig4, max_block_size;
+  uint32_t sig4, max_block_size, num_rows, row_offset;
```
Please initialize to zero if there's no way to declare at the point of initialization.

vuule commented on cpp/src/io/avro/avro.cpp:
```diff
@@ -67,8 +67,10 @@ std::string container::get_encoded()
-  uint32_t sig4, max_block_size;
+  uint32_t sig4, max_block_size, num_rows, row_offset;
   size_t total_object_count;
+  uint64_t sync_marker[2];
+  bool valid_sync_markers;
```
Looks like these can have a scope limited to the block in which they are used.

vuule reviewed (commented):
Great PR, really appreciate the detailed comments!
Got a few more nitpicks, and I think they all stem from consistency with current code (which is in of clean up) :D

wence- reviewed (approved):
Thanks! Python changes look great.

tpn commented:
Just pushed a rebase w/ latest changes + the "drop dead GPU decoding logic" commit dropped -- turns out that path *is* actually hit in some of the test single value tests.

vuule commented:
/ok to test

vuule commented:
> Just pushed a rebase w/ latest changes + the "drop dead GPU decoding logic" commit dropped -- turns out that path _is_ actually hit in some of the test single value tests.

Sounds like I'm missing something in the logic, will need to take a look. Thanks for checking.

vuule reviewed (approved):
:fire:

ttnghia reviewed (approved):
Thank you so much for the documentation. That is fantastic! I really appreciate it.

tpn commented:
> Thank you so much for the documentation. That is fantastic! I really appreciate it.

Thanks for such responsive and constructive PR feedback!  Makes the process go a lot smoother :-)

tpn commented:
Just pushed an up-to-date-with-23.06 rebase.  Need anything else from me at this point?  We're just waiting for another review on your side, right?

vuule commented:
/ok to test

vuule commented:
> Just pushed an up-to-date-with-23.06 rebase. Need anything else from me at this point? We're just waiting for another review on your side, right?

There's two C++ reviews and a Python review, so the PR is good to go as soon as CI passes.
